contracts/PionexContract.sol

```diff
@@ -0,0 +1,367 @@
+        address _weth,
+        address _permStorage,
+        address _spender,
+        address _coordinator,
+        uint256 _factorActivateDelay,
+        address _feeCollector
+    ) StrategyBase(_owner, _userProxy, _weth, _permStorage, _spender) {
+        coordinator = _coordinator;
+        factorActivateDelay = _factorActivateDelay;
+        feeCollector = _feeCollector;
+    }
+
+    receive() external payable {}
+
+    /// @notice Only owner can call
+    /// @param _newCoordinator The new address of coordinator
+    function upgradeCoordinator(address _newCoordinator) external onlyOwner {
+        require(_newCoordinator != address(0), "PionexContract: coordinator can not be zero address");
+        coordinator = _newCoordinator;
+
+        emit UpgradeCoordinator(_newCoordinator);
+    }
+
+    /// @notice Only owner can call
+    /// @param _tokenlonFeeFactor The new fee factor for user
+    function setFactors(uint16 _tokenlonFeeFactor) external onlyOwner {
+        require(_tokenlonFeeFactor <= LibConstant.BPS_MAX, "PionexContract: Invalid user fee factor");
+
+        pendingTokenlonFeeFactor = _tokenlonFeeFactor;
+
+        factorsTimeLock = block.timestamp + factorActivateDelay;
+    }
+
+    /// @notice Only owner can call
+    function activateFactors() external onlyOwner {
+        require(factorsTimeLock != 0, "PionexContract: no pending fee factors");
+        require(block.timestamp >= factorsTimeLock, "PionexContract: fee factors timelocked");
+        factorsTimeLock = 0;
+        tokenlonFeeFactor = pendingTokenlonFeeFactor;
+        pendingTokenlonFeeFactor = 0;
+
+        emit FactorsUpdated(tokenlonFeeFactor);
+    }
+
+    /// @notice Only owner can call
+    /// @param _newFeeCollector The new address of fee collector
+    function setFeeCollector(address _newFeeCollector) external onlyOwner {
+        require(_newFeeCollector != address(0), "PionexContract: fee collector can not be zero address");
+        feeCollector = _newFeeCollector;
+
+        emit SetFeeCollector(_newFeeCollector);
+    }
+
+    /// @inheritdoc IPionexContract
+    function fillLimitOrder(
+        PionexContractLibEIP712.Order calldata _order,
+        bytes calldata _orderUserSig,
+        TraderParams calldata _params,
+        CoordinatorParams calldata _crdParams
+    ) external override onlyUserProxy nonReentrant returns (uint256, uint256) {
+        bytes32 orderHash = getEIP712Hash(PionexContractLibEIP712._getOrderStructHash(_order));
+
+        _validateOrder(_order, orderHash, _orderUserSig);
+        bytes32 allowFillHash = _validateFillPermission(orderHash, _params.pionexTokenAmount, _params.pionex, _crdParams);
+        _validateOrderTaker(_order, _params.pionex);
+
+        // Check provided pionexToken/userToken ratio is better than or equal to user's specfied pionexToken/userToken ratio
+        // -> _params.pionexTokenAmount/_params.userTokenAmount >= _order.pionexTokenAmount/_order.userTokenAmount
+        require(
```

Comment:
Since the `minPionexTokenAmount` should be checked after fee deducted, may be this require can be removed since it'd be reverted if `pionexTokenForUser < _settlement.minPionexTokenAmount` even the ratio here is better?

Reply:
Good call. Fixed in ed48d9b